### DISCUSSION TITLE
multitenancy with  beyondcode / laravel-websockets

### DISCUSSION THREAD

**Discussion by sys-auditing** [spatie/laravel-multitenancy#316] [Q&A]
Hi,

how can i integrate your package with `laravel-websockets`, noting that i'm using `multi databases` approach?

Any help will be appreciated.

Thanks

**Comment by masterix21** [comment]
You have two ways:
- an app per tenant on you websockets installation
- one app for all tenants, prefixing the channels like so `tenants.{tenantId}.{yourChannel}`

**Reply by sys-auditing** [reply]
Thank you @masterix21,

can you give me a simple example for each approach, because I'm having difficulty understanding how to make it work..

Should i make a migration(websockets statistics table) in landlord DB?
For the second method, can i choose an app from websockets statistics dashboard?

Thank you

**Reply by masterix21** [reply]
I have no example code to share with you. So it's easy to implement it.

For the first suggestion, you can read the laravel-websockets docs.

For the second suggestion, you can read how to broadcast a notification in the Laravel documentation.

**Reply by sys-auditing** [reply]
yes, i see.

I have an other issue : 

//ActionWasFired event 

```PHP
class ActionWasFired implements ShouldBroadcast
{
    use Dispatchable, InteractsWithSockets, SerializesModels;

    private $user;
    private $tenant;
    /**
     * Create a new event instance.
     *
     * @return void
     */
    public function __construct($user, $tenant)
    {
	    $this->user   = $user;
	    $this->tenant = $tenant;
    }   
    
    public function broadcastQueue()
    {
        return 'broadcastable';
    }

    public function broadcastAs()
    {
        return 'critical-action';
    }

    public function broadcastWith()
    {
        return [
            'user'       => $this->user->id,
            'tenant'     => $this->tenant->id,
        ];
    } 

    /**
     * Get the channels the event should broadcast on.
     *
     * @return \Illuminate\Broadcasting\Channel|array
     */
    public function broadcastOn()
    {
        return new PrivateChannel(
            "actions.{$this->tenant->id}.{$this->user->id}"
        );
    }
}
```
//channels.php
```PHP
Broadcast::channel('actions.{tenant_id}.{id}', function ($user, $tenant_id, $id) {
    return ((int) $user->id === (int) $id && (int) app('currentTenant')->id === (int) $tenant_id);
});
```
//BroadcastServiceProvider.php
```PHP
public function boot()
{
     Broadcast::routes([ "middleware" => ["web", "auth", "tenant"] ]);
     require base_path('routes/channels.php');
 }
```
When firing event via tinker using :
```PHP
Tenant::first()->makeCurrent();
$user = User::find(1);
event(new App\Events\ActionWasFired($user, app('currentTenant')));
```
all works as a charm, but when calling this event from a controller i got this error message : 
```
local.ERROR: SQLSTATE[3D000]: Invalid catalog name: 1046 No database selected (SQL: select * from `users` where `users`.`id` = 1 limit 1) 
```

If you could point me in the right direction, that would be great.

**Reply by sys-auditing** [reply]
## UPDATE
### Summarized solution

The event in question should implementing ``` TenantAware ``` interface.
Thank you
